### ISSUE TITLE
Vaccination Certificate disappears from UI

### ISSUE THREAD

**Issue by vaubaehn** [corona-warn-app/cwa-app-android#3549]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [x] Bug is not already reported in another issue

## Technical details

- Device name: unknown, probably several
- Android version: unknown
- App version: 2.3.4

## Describe the bug
Since vaccination DCCs can be imported into CWA, I found several reviews in Google Play Store, that the certificate "vanishes" by itself. According to the following report, it's still persisted, but not displayed:
> Ich muss mein impfzertifikat täglich neu hinzufügen, da es einfach verschwindet... Nehm ich den digitalen impfausweiss halt lieber doch analog mit... Update: ich bin kein Idiot. Das Zertifikat fehlt. Aber Plot Twist: wenn ich es nochmal hinzufügen will kommt das es schon da ist. Ist also ein reiner GUI Glitch...

## Steps to reproduce the issue

1. Import/scan vac DCC
2. Wait some time (till next day), find it vanished
3. Try to re-import DCC
4. See error: Already scanned.

## Expected behaviour

The DCC should be displayed always after it was persisted.

## Additional context

I hope for your understanding, that I create an issue for a bug that I don't experience myself.
But as I already found quite a few reports of this kind, I thought could be good to track it here. Please have a look to Play Store reviews yourself. Maybe that issue is already known to the dev team, but then we have a place to track progress from here.

Edit: More reports
![Screenshot_2021-06-25-10-07-32~2](https://user-images.githubusercontent.com/68376758/123392659-9d441a80-d59d-11eb-9d55-96c86d2c8acc.png)
![Screenshot_2021-06-25-10-14-19~2](https://user-images.githubusercontent.com/68376758/123393304-3ecb6c00-d59e-11eb-8fd2-3418f8b0de95.png)
![Screenshot_2021-06-25-10-19-33~2](https://user-images.githubusercontent.com/68376758/123394034-fb253200-d59e-11eb-8ff9-763032ae9caa.png)
![Screenshot_2021-06-25-11-11-44~2](https://user-images.githubusercontent.com/68376758/123401108-c0bf9300-d5a6-11eb-8ca3-58f6cf5ed1b1.png)

---
Internal Tracking ID: [EXPOSUREAPP-8117](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8117) [Android] Vaccination certificate suddenly disappears
Internal Tracking ID: [EXPOSUREAPP-8113](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8113) Android, after update to 2.4
Internal Tracking ID: [EXPOSUREAPP-8114](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8114) Ios, after update to 2.4

**Comment by Shagonaut** [comment]
iOS here. I have the same problem.

**Comment by Ein-Tim** [comment]
@Shagonaut 

Did you update to version 2.4? The certificate has moved to it's own tap, "Certificates".

**Comment by Shagonaut** [comment]
@Ein-Tim Thanks for your response! Had no automatic update or information in the overview. So I checked manually and yes, it weren't updated to 2.4. - Updated it and now the certificate is visible again. Thank you!

**Comment by vaubaehn** [comment]
@Shagonaut 

> Updated it and now the certificate is visible again.

So, there was a problem with iOS also... Interesting. For Android I didn't find any bug fix or similar for version 2.4.x. So, please report back here in case your certificate vanishes again.

Thank you!

**Comment by Ein-Tim** [comment]
@vaubaehn 

I assume there was no bug but it's only bad UX that the register your certificate box is still shown even after you have already added your certificate. 
This is(was) quite confusing.

**Comment by vaubaehn** [comment]
@Ein-Tim 
I was also thinking about this, but if you read above reports carefully, it does not seem to be the case (mostly the first one 😉 )
It does also not fit to what @Shagonaut reported, that after the update the DCC was displayed again

**Comment by Ein-Tim** [comment]
@Shagonaut 

Are you sure that under the risk card there was no certificate shown? Or did you only see the "Register your certificate" box?

**Comment by vaubaehn** [comment]
@Ein-Tim 
And if people report, for 2 days everything was fine and then they were gone, then there was "a change". That smells more like a bug, rather than problems with biological memory.

**Comment by MikeMcC399** [comment]
@vaubaehn 
> That smells more like a technical bug,

That is my feeling too, because the "Digital Proof of Vaccination" tile (if shown) is directly below the risk tile and above the "Register Your Test" tile so it would be hard to overlook.

**Comment by Ein-Tim** [comment]
> That smells more like a technical bug, rather than problems with biological memory.

At least under iOS I disagree and. Just take a look [here](https://github.com/corona-warn-app/cwa-app-ios/issues/2969#issuecomment-863944476) where a user didn't found the vaccination certificate but only the box to register a new one. It was bad UX, at least under iOS, and I will later provide a screen recoding.

**Comment by vaubaehn** [comment]
> At least under iOS I disagree. Just take a look [here](https://github.com/corona-warn-app/cwa-app-ios/issues/2969#issuecomment-863944476) where a user didn't found the vaccination certificate but only the box to register a new one. It was bad UX, and I will later provide a screen recoding which shows this.

But even for iOS I understood it to be a bug of not updating the home screen, thus not showing the DCC card at all. It needed a "hard close"/force stop and restarting of CWA for some users for the card to show up again, see here:
https://github.com/corona-warn-app/cwa-app-ios/issues/2969#issuecomment-863852827
and also for the user you referenced, he did the same:
https://github.com/corona-warn-app/cwa-app-ios/issues/2969#issuecomment-863897684
and in both cases it took a while until the card showed up... That's how I understand it 🤔 

Updating CWA like @Shagonaut did has the same effect.
But I thought we'd try to figure out, what's going wrong on Android? 😄 
Could both OS issues really be connected?

**Comment by Ein-Tim** [comment]
> But I thought we'd try to figure out, what's going wrong on Android? 😄
Could both OS issues really be connected?

No. And even if, under iOS it seems to be fixed in version 2.4. So it's only about Android and whether version 2.4 still has this issue.

**Comment by vaubaehn** [comment]
> No. And even if, under iOS it seems to be fixed in version 2.4

Can you reference a PR? That could help. Or do you assume this because of @Shagonaut 's experience?

**Comment by Ein-Tim** [comment]
@vaubaehn I assume this because the logic has changed with the move of the certificate to it's own tab.

**Comment by vaubaehn** [comment]
@mtwalli @harambasicluka 
This issue is originally about DCC card "sometimes disappearing from home screen" with CWA 2.3.4

I'm assuming that
- the issue reported here is indeed a bug,
- probably Android specific,
- occurs only occasionally and in rare cases (given the number of reports),
- and could have a root cause that also may have an effect on other issues that have been reported here and are hard to reproduce.

I'm trowing my dart towards two different theories, what could have happend:

**Theory A**
Some data that the UI relies on (or even UI states) is intentionally **or accidentally** stored in Android's cache, i.e. inside the CWA-cache-directory. When Android OS decides to clean up the cache automatically (due to lack of free memory) the data is lost.
I stumbled across this at first, after my CWA crashed from time to time, and my activated debug log feature occasionally reset itself. Discussing this issue with @d4rken (DON'T READ HERE IF YOU'RE ON VACATION), it turned out that debug log stores by design the log inside the cache directory. And because my family-in-law loves to send cat videos via whatsapp, my old phone's memory has a hard time, resulting in automatic cache sweeps, and in this case loss of debug log. If other CWA data is stored there, for some CWA users it may have unexpected results when the cache is cleared.
Accidentally storing data inside cache could be due to a wrong storing path constant, or chosing the wrong interface/class for storing in some place.

**Theory B**
A worker is used to process and/or store data. While the worker is still initializing, data is already updated in the process flow and a new worker is going to be started for processing/storing the updated data. As same class workers are blocked as siblings and cancelled, this _may_ result in data loss for the updated data, in case this special (race) condition is not handled gracefully (which is not trivial at all, I guess).
This actually happens. While recently checking my debug logs, I found several entries where workers have been cancelled due to siblings (a probably related issue: https://github.com/corona-warn-app/cwa-app-android/issues/2880#issuecomment-832166500, find associated logcat here: https://github.com/corona-warn-app/cwa-app-android/issues/2880#issuecomment-827846585). The maximum number of siblings was always 1. But most of the time, no siblings were existing. This is why I assume, this may mostly affect old and slow devices under high work load, when workers cannot be started fast enough.

In some ways these unexpected behaviours of "disappearing data" _may_ also fit to following issues:
https://github.com/corona-warn-app/cwa-app-android/issues/2986 (event not persisted)
https://github.com/corona-warn-app/cwa-app-android/issues/2880 (possible worker concurrency/race condition?)
https://github.com/corona-warn-app/cwa-app-android/issues/3093 (UI state "lost"/not persisted? vs. ENS API timeout)

A short feedback would be appreciated, thank you.

**Comment by vaubaehn** [comment]
Now after people updated to 2.4.x (consciously manually or unconsciously automatically) there are new reports of problems:
- missing Vac-DCC - that does not seem to be related with new UI certificates tab, as person tried to re-scan QR, which is obviously only possible from certificates tab (see screenshot), so it's obviously stored but not displayed.
- A 'full' valid vaccination turned to 'valid in 0 days'  which looks like data corruption.

It's confusing, that only some people are affected. Either their DCC JSON is not fully compatible with schema and the adjustments for validation of DCC implemented into 2.4.x broke it, or there is indeed a problem with storage (see above).

<img src="https://user-images.githubusercontent.com/68376758/123516640-6fde9600-d69d-11eb-9eb9-77e0623ca985.png" width="200">
<img src="https://user-images.githubusercontent.com/68376758/123516672-98ff2680-d69d-11eb-9903-0d13857f877e.png" width="200">

**Comment by Ein-Tim** [comment]
> A 'full' valid vaccination turned to 'valid in 0 days' which looks like data corruption.

For me this sounds like a vaccination certificate which was shown as "fully" valid on day 14 after the last dose already turned into the certificate that is valid in 0 days. 
This would be  a side effect of the fix for https://github.com/corona-warn-app/cwa-documentation/issues/638

> In general I'm voting for better UX in future, in case cards are moved from home screen into a different tab or a menu

Could you maybe open a wishlist item for this, please? I fully support this!

**Comment by DerVogel2020** [comment]
> In general I'm voting for better UX in future, in case cards are moved from home screen into a different tab or a menu (e.g., risk card, test registration, statistics, event creation... The card that is moved could be replaced by a placeholder, that informs user that the function has been moved and where to find it. After tapping on it once, it could be dismissed/removed.

I also missed my certificate after updating.  A hint where to find it now would have been nice.

For a wishlist entry it's IMHO too late now, since the vaccination certificate now have been moved to the new tab.

**Comment by Ein-Tim** [comment]
@DerVogel2020 

Yes, for the vaccination certificate you're right, thats too late now. But they did the same with the contact diary back then, so maybe a more general wishlist issue requesting better information if there is an UI change?

**Comment by MDXDave** [comment]
In the app Version 2.4.2 I am not able to add a New vaccination certificate. 
The certificate tab says that I can add a New certificate on the home screen - but there is no card that let me add a new certificate (see screenshots). 

![Screenshot_20210627-195106~2](https://user-images.githubusercontent.com/6099976/123554732-b5748f00-d781-11eb-8118-2487d4d272a4.jpg)
![Screenshot_20210627-195111](https://user-images.githubusercontent.com/6099976/123554733-b6a5bc00-d781-11eb-9ed8-255e3552d2c2.jpg)

**Comment by Ein-Tim** [comment]
@MDXDave 

Please click on the vaccination certificate which you have already added, there you can add the one you got with your second dose.

You can register tests on the home screen, for which you now can request a test certificate.

**Comment by MDXDave** [comment]
Sorry, I thought it should be able to register certificates for more than one person already. 😄

**Comment by d4rken** [comment]
We need to take care not to mix 3 different issues.
"cert disappearing" is just a symptom, possible causes seems to be:

1. UI Change (not an error?)
2. Missing but can't be readded (UI error?)
3. Missing and can be added again (storage error?)

I'm struggling to narrow this down.
Would be great if someone could provide a debug log.
At least case 2 seems to be a continuous situation that should make a debug log possible.

**Comment by d4rken** [comment]
TL;DR Can't reproduce yet, still looking into it.

Writing things down helps me think through stuff, thought it might more interesting to put it here than on an internal ticket.

* I'm not 100% sure whether Cause#2 isn't actually Cause#1. 
* Wondering whether the missing certs will pop up with 2.5 due to family cert support, not sure if anyone still has the bug though or everyone reset?
* If i set `chattr -i vaccination_localdata.xml` (write protect the data), then I can reproduce 3 in a way that some users described. SharedPreference write to memory worked, write to disk failed, data lost when app session is restarted.
    * Could this be a low storage issue? No user mentioned anything about low storage though.
    * If it is not low storage, could it be a file system issue? But I would expect that to affect more than just the certs...
    * Was something like "adopt sdcard as internal storage"  used?
    * It's a pity that SharedPreferences are not very error throwing happy

https://github.com/corona-warn-app/cwa-app-android/issues/3549#issuecomment-868530829

> Theory A

Couldn't find anything like this (cache theory).

> Theory B

Plausible, but I find it unlikely. I don't think workers are the issue here. We don't use workers for immediate in-app action, just reliable background processing.
Scanning a certificate moves it immediately into the repository.
If the user saw the certificate, that has happened.

https://github.com/corona-warn-app/cwa-app-android/blob/25403e0e90c2ca1534a198a84380afdbb65c7672/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/VaccinationRepository.kt#L44

It's then part of the `HotDataFlow` which is designed to handle the race-conditions while the order can be jumbled, it only uses functions to update the data so we never overwrite data in a race condition `new = function(old)`

Storing to disk happens async though:
https://github.com/corona-warn-app/cwa-app-android/blob/25403e0e90c2ca1534a198a84380afdbb65c7672/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/VaccinationRepository.kt#L62-L74

It just listens to every change that happens to the in memory data and saves it. I don't see how a racecondition could break this.
It could break and produce the described user behavior if we don't write the data or writing it fails.
Not writing would mean that the user kills the app session, or the app crashes (no user mention?) before data is written.
But that would only be reasonable if the write operation is delayed long enough for the app session to die (a write takes miliseconds). This could happen all threads in coroutine context are in use (the coroutine is starved). If we had used `Dispatcher.Default` I maybe could see that happen on low end dual core devices? (`Default` is ~ amount of cores), but the writer here uses the `IO` dispatcher which will just create new threads if needed.

This leaves an error during writing the data to storage. Any error would crash the app though (re `throw` in the `catch` clause and we are on `AppScope`).
Some research shows that `SharedPreferences` doesn't really throw many exceptions though...
Throwing an error here would be nice, but an error, so low level in Android code, not sure what the app could do about that.

> In some ways these unexpected behaviours of "disappearing data" may also fit to following issues:
> #2986 (event not persisted)

Also strange but different storage mechanisms, checkins use SQL database, SQL database operations are also not shy about throwing errors on low storage. Will write down to check error handling there though, maybe there was an error but it wasn't surfaced.

> #2880 (possible worker concurrency/race condition?)

Still unclear on causes, don't think it's related.

> #3093 (UI state "lost"/not persisted? vs. ENS API timeout)

Not related, afaict mostly bad UI behavior due ENS API calls taking minutes instead of miliseconds. Not sure on cause yet. Seems  also in the area of weird ENS behavior, a bit like #2880?

**Comment by harambasicluka** [comment]
PR: https://github.com/corona-warn-app/cwa-app-android/pull/3742/files

**Comment by Ein-Tim** [comment]
Here's a Twitter user reporting this behavior: https://twitter.com/Pterodactylus2/status/1420461784314564613?s=20

**Comment by dsarkar** [comment]
@Ein-Tim Thanks. Information will be forwarded.

**Comment by heinezen** [comment]
Our official recommendation is to restart the device if something like this happens.

----
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
@heinezen The Twitter user [tried](https://twitter.com/pterodactylus2/status/1421112951155699723?s=21) a restart and it did not help.

**Comment by mtwalli** [comment]
@Ein-Tim  can we get a log report from any of the users by any chance?

**Comment by vaubaehn** [comment]
@mtwalli @d4rken 
I'm beginning to wonder, if the setup/combination of all components may lead to problems with Kotlin's `flow`, resulting in problems to receive the `emits ` in some cases... Most likely not exactly like [the issue described here](https://github.com/Kotlin/kotlinx.coroutines/issues/2835), but something kind of... I remember the problems with R8 from last year (app crash just for a few java implementations)... Maybe the issue here is a similar edge case?

> @Ein-Tim can we get a log report from any of the users by any chance?

Maybe @dsarkar and @heinezen can ask the Android Play Store team to get more affected users for a log report?
But according to most of the reports the issue occurs _after_ any update.  So users would need to enable error report _before_ the next update and send the report after...

**Comment by d4rken** [comment]
> I'm beginning to wonder, if the setup/combination of all components may lead to problems with Kotlin's `flow`, resulting in problems to receive the `emits ` in some cases... Most likely not exactly like [the issue described here](https://github.com/Kotlin/kotlinx.coroutines/issues/2835), but something kind of... I remember the problems with R8 from last year (app crash just for a few java implementations)... Maybe the issue here is a similar edge case?

The delivered APK (including code shrinking) is the same for all users. If such an R8 related issue would exist, it would affect all users. It's likely a racecondition. Less likely but plausible an edge case specific to the users data. But all data paths I checked, would not have silently thrown the data away when this happen, they would have crashed the app.

> Maybe @dsarkar and @heinezen can ask the Android Play Store team to get more affected users for a log report?
> But according to most of the reports the issue occurs _after_ any update. So users would need to enable error report _before_ the next update and send the report after...
> Edit: Forget about the last sentence above. A single point error log after the issue occured would already help to understand, what CWA is "seeing" when trying to map the certificates.

Users on Google Play have already been asked, before the update rollout.

I don't think it's an update issue, the update is just a guaranteed app restart after which there is a chance again to reach the problematic state.

An error log from someone who doesn't see the vaccination certificate, but also gets an error message when trying to scan it again, would help the most to further debug this.

@Ein-Tim's Twitter user is still on 2.5.1, it's possible that the performance improvements in 2.6.x successfully mitigate the issue 🤞 .

**Comment by vaubaehn** [comment]
Hi @d4rken , thanks a lot for your thorough replies!
There is not much more to say (than thank you for being on), just sharing some final(?) thoughts:

> We need to take care not to mix 3 different issues.
> "cert disappearing" is just a symptom, possible causes seems to be:
>   1.  UI Change (not an error?)
>   2.  Missing but can't be readded (UI error?)
>   3.  Missing and can be added again (storage error?)
>    I'm not 100% sure whether Cause#2 isn't actually Cause#1.
>    Was something like "adopt sdcard as internal storage"  used?

From all reviews in Google Play Store, there were indeed many cause#2 is actually cause#1. But there were obviously also distinct cause#2 cases. I guess cause#3 is quite unlikely, except for users with "throw-everything-on-external-SD"-OS...

> > Theory A
> 
> Couldn't find anything like this (cache theory).

I explicitly tried to wipe cache - but all certificates stayed where they should.

> It's then part of the `HotDataFlow` which is designed to handle the race-conditions while the order can be jumbled, it only uses functions to update the data so we never overwrite data in a race condition `new = function(old)`
> 
> Storing to disk happens async though:
> 
> https://github.com/corona-warn-app/cwa-app-android/blob/25403e0e90c2ca1534a198a84380afdbb65c7672/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/VaccinationRepository.kt#L62-L74
> 
> It just listens to every change that happens to the in memory data and saves it. I don't see how a racecondition could break this.
> It could break and produce the described user behavior if we don't write the data or writing it fails.
> Not writing would mean that the user kills the app session, or the app crashes (no user mention?) before data is written.
> But that would only be reasonable if the write operation is delayed long enough for the app session to die (a write takes miliseconds). This could happen all threads in coroutine context are in use (the coroutine is starved). If we had used `Dispatcher.Default` I maybe could see that happen on low end dual core devices? (`Default` is ~ amount of cores), but the writer here uses the `IO` dispatcher which will just create new threads if needed.

Hmmm... `HotDataFlow` + `IO-Dispatcher` + `R8 minifiedEnabled = true` => https://github.com/Kotlin/kotlinx.coroutines/issues/2835

But then:
> The delivered APK (including code shrinking) is the same for all users. If such an R8 related issue would exist, it would affect all users

The reason why I thought, that such an issue has a chance to only affect a limited number of users was this:
https://github.com/corona-warn-app/cwa-app-android/issues/1609
What if the compiled byte code here does not result in an app crash but in a NOP for their VMs? Maybe we should ask affected users for their chipset? 😉 
But right, this time it's not limited to Android 7 and thus unlikely the cause of this issue.

> It's likely a racecondition.

I hope you win the race! 🏎️ 🤞 

Have a nice week! 🍻

**Comment by Ein-Tim** [comment]
> @Ein-Tim's Twitter user is still on 2.5.1, it's possible that the performance improvements in 2.6.x successfully mitigate the issue 🤞 .

@d4rken 

I [asked](https://twitter.com/eintim2/status/1424118419364921346?s=21) the Twitter user and she said that the certificate was still not there. 

---

> @Ein-Tim can we get a log report from any of the users by any chance?

@mtwalli 

I can ask them! Please let me know in which situation exactly they should start logging and what they are supposed to do while the error log is running.

**Comment by vaubaehn** [comment]
Hi @mtwalli ,

recent comment from Matthias (who may be on well-deserved vacation?) above:
> An error log from someone who doesn't see the vaccination certificate, but also gets an error message when trying to scan it again, would help the most to further debug this.

I found [that one on Google Play Store](https://play.google.com/store/apps/details?id=de.rki.coronawarnapp&reviewId=gp%3AAOqpTOGWl9Ugt_YEZazPK0hF5sTzx11u6TfI2wDMgxdYMKChNCSerAHqWOuYCKCKqH_f44VOPVoi_6PEf005lw):
ID D2EC482DDF153608FCE8

Just in case, it's still useful. Report was on August 3, so only some days left to check it.

Cheers,
v.

**Comment by vaubaehn** [comment]
@Ein-Tim 

> I can ask them! Please let me know in which situation exactly they should start logging and what they are supposed to do while the error log is running.

Until we don't hear anything different from @mtwalli :
1. Start error log any time
2. Go to certificate screen
3. See no certificates
4. Try to re-scan a certificate that was previously scanned
5. See error message
6. Maybe: close the app ("swipe app away" (depending on OS/MUI) or from app selector (rectangle) in Android navigation), then restart the app and repeat steps 2 to 5
7. send error log

Release 2.7 will get extended error logging to hunt the bug: PR https://github.com/corona-warn-app/cwa-app-android/pull/3901
fingers crossed! 🤞

**Comment by Ein-Tim** [comment]
@vaubaehn Okay, thank you! I'll try to get in touch with the user and let you know when I have news.

**Comment by dsarkar** [comment]
@vaubaehn Thanks for mentioning the error ID. This has been added to the internal ticket some days ago. However, we will ping some devs again.

**Comment by Ein-Tim** [comment]
@vaubaehn 

Bad news, the Twitter user [did not keep](https://twitter.com/pterodactylus2/status/1425719361218695171?s=21) the sheet of paper with the QR-Code on it, so there is no way for her to rescan the certificate... 😕

**Comment by FloSchwalm** [comment]
I just experienced this issue. In my case the cause seemed to be Lie-fi - phone was connected to cell, but no data actually getting through. The workaround in my case was to activate airplane mode and restart the app, then the certificate was displayed again. Is there some kind of data update happening that prevents the UI from displaying the certificate while waiting for a response?

**Comment by vaubaehn** [comment]
@FloSchwalm @mtwalli (and @d4rken 😢 )

> I just experienced this issue. In my case the cause seemed to be Lie-fi - phone was connected to cell, but no data actually getting through. The workaround in my case was to activate airplane mode and restart the app, then the certificate was displayed again. Is there some kind of data update happening that prevents the UI from displaying the certificate while waiting for a response?

Looks like the most promising feedback until now:
The value sets are part of the `flow` . What, if there was a connection problem (switching/loosing network, timeout...) when polling for value sets? Can we be sure that this condition is handled properly while being online? Or can it get stuck somewhere (error not handled as expected), and emitting of value set is blocked somehow? Re-starting the app like @FloSchwalm did could handle any race condition (which should not occur in `flow`). But being in airplane-mode could CWA force to pull an already stored value set...

**Comment by dsarkar** [comment]
@vaubaehn @FloSchwalm Thanks for discussing here: Forwarded to internal ticket, pinged dev team.

**Comment by d4rken** [comment]
@FloSchwalm's issue has the same symptoms but maybe not the same cause. You are on v2.7.x? :thinking: 

I looked into the network situation. We can simulate a stuck network situation by adding an interceptor within the client created by `HttpModule` or just do some firewall shenanigans.

### On CWA v2.6.x
The certificates load instant as no network call is involved. The only delay is local processing/rule-checking, nothing significant enough to be called "hanging" (and the debug build is slower than prod anyways). I also artificially restricted the thread pool to see if a hanging network check could starve other coroutines, but the certificates still load quickly.

### On CWA v2.7.x
**If the locally cached config is outdated**, the certificates tab may be empty for up to 10 seconds.
This is due to the state being checked for each emission and the state check requiring app-config data:
https://github.com/corona-warn-app/cwa-app-android/blob/97fbc95b71db96ed637af3eaa35118eb705a1611/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/statecheck/DccStateChecker.kt#L28

Even though `DccStateChecker` only requests the `currentConfig` and doesn't force a check with `getAppConfig()`, it will still hang if the app was just started and it is the first time the config is accessed.

[`AppConfigProvider` ](https://github.com/corona-warn-app/cwa-app-android/blob/e5eb5f483675e7c8475538bd24a64b7329cbe872/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt#L28), when initializing, calls [`AppConfigSource.getConfigData()`](
https://github.com/corona-warn-app/cwa-app-android/blob/e5eb5f483675e7c8475538bd24a64b7329cbe872/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/internal/AppConfigSource.kt#L32-L35) which will perform a network request if the local config is no longer valid.

So if the app was no longer in memory and the user opens the app (ignores the empty home screen feed which is also waiting for app config data), then the certificates tab will be empty until `AppConfigProvider` has it's initial value, which in  case of bad network would be after up to 10 seconds (timeout). Anything that depends on app-config will be blocked until that initial value is available. 

I would probably change `AppConfigSource.getConfigData()` to something like `getConfigData(offlineOnly: Boolean = false)` so that it skips the remote config request if set, and then use that to init `AppConfigProvider`.

Any part of the app that requires the absolute latest config uses `AppConfigProvider.getAppConfig()` AFAIK, and all components that are fine with the `current` one, there is no longer the chance to hang in this edge case.

So maybe the issue from 2.3 to 2.5 (UI confusion aside) was fixed with 2.6.x, and this is just the same symptom with a new cause?

**Comment by FloSchwalm** [comment]
Yes, my App version is 2.7.1
Your description definitely fits what I experienced. It's probably possible that I closed the app and moved on to airplane mode before the 10-19 seconds passed as I was supposed to present my certificate and was in a hurry not to hold up the line longer than necessary.

**Comment by vaubaehn** [comment]
Still a problem after updates for some users, e.g.,
https://play.google.com/store/apps/details?id=de.rki.coronawarnapp&reviewId=gp%3AAOqpTOFG0hgrr7ZDtuVbeGY2KU6ZnmSMIZRy9-Ctk98qS46ZyC8LJd39OVKWCt0w5_9KWOQ_6dA99jpmM-qGCA

**Comment by svengabr** [comment]
I have just checked the linked Jira issue.

Jira Ticket is flagged as:
Resolution: `Wont Fix`
Status: `Obsolete`

Developer comment:
> Setting to obsolete, duplicate.
>
> Continue in EXPOSUREAPP-8113

**Comment by Ein-Tim** [comment]
@svengabr 

Whats the status of EXPOSUREAPP-8113?

**Comment by Ein-Tim** [comment]
@svengabr Ok, your call whether to leave this issue closed or reopen it.
